src/wallet.rs

```diff
@@ -233,46 +237,115 @@ where
 		self.get_balances(total_anchor_channels_reserve_sats).map(|(_, s)| s)
 	}
 
-	/// Send funds to the given address.
-	///
-	/// If `amount_msat_or_drain` is `None` the wallet will be drained, i.e., all available funds will be
-	/// spent.
 	pub(crate) fn send_to_address(
-		&self, address: &bitcoin::Address, amount_msat_or_drain: Option<u64>,
+		&self, address: &bitcoin::Address, send_amount: OnchainSendType,
 	) -> Result<Txid, Error> {
-		let confirmation_target = ConfirmationTarget::OutputSpendingFee;
-		let fee_rate = FeeRate::from_sat_per_kwu(
-			self.fee_estimator.get_est_sat_per_1000_weight(confirmation_target) as f32,
-		);
+		let confirmation_target = ConfirmationTarget::OnchainPayment;
+		let fee_rate = self.fee_estimator.estimate_fee_rate(confirmation_target);
 
 		let tx = {
 			let locked_wallet = self.inner.lock().unwrap();
 			let mut tx_builder = locked_wallet.build_tx();
 
-			if let Some(amount_sats) = amount_msat_or_drain {
-				tx_builder
-					.add_recipient(address.script_pubkey(), amount_sats)
-					.fee_rate(fee_rate)
-					.enable_rbf();
-			} else {
-				tx_builder
-					.drain_wallet()
-					.drain_to(address.script_pubkey())
-					.fee_rate(fee_rate)
-					.enable_rbf();
+			// Prepare the tx_builder. We properly check the reserve requirements (again) further down.
+			match send_amount {
+				OnchainSendType::SendRetainingReserve { amount_sats, .. } => {
+					tx_builder
+						.add_recipient(address.script_pubkey(), amount_sats)
+						.fee_rate(fee_rate)
+						.enable_rbf();
+				},
+				OnchainSendType::SendAllRetainingReserve { cur_anchor_reserve_sats } => {
+					let spendable_amount_sats =
+						self.get_spendable_amount_sats(cur_anchor_reserve_sats).unwrap_or(0);
+					let mut tmp_tx_builder = locked_wallet.build_tx();
+					tmp_tx_builder
+						.add_recipient(address.script_pubkey(), spendable_amount_sats)
+						.fee_rate(fee_rate)
+						.enable_rbf();
```

Comment:
Couldn't this result in a transaction where the entire `cur_anchor_reserve_sats` goes to fees? And thus there could be one less output, which would cause the fee estimation to be off?

Would the following solve this?

```rust
let change_address = locked_wallet.get_internal_address(AddressIndex::Peek(0));
tmp_tx_builder
	.drain_wallet()
	.drain_to(address.script_pubkey())
	.add_recipient(change_address, cur_anchor_reserve_sats)
	.fee_rate(fee_rate)
	.enable_rbf();
```

Though if `cur_anchor_reserve_sats` was exactly covered by one utxo, then an extra input would be used in the estimation. 🤔